Lib/unittest/test/testmock/testasync.py

```diff
@@ -528,11 +528,158 @@ class AsyncMockAssert(unittest.TestCase):
+
+        with self.assertRaises(AssertionError):
+            self.mock.assert_called()
+
+        asyncio.run(self._runnable_test())
+        self.mock.assert_called_once()
+        self.mock.assert_awaited_once()
+
+    def test_assert_called_twice_and_awaited_once(self):
+        mock = AsyncMock(AsyncClass)
+        coroutine = mock.async_method()
+        with self.assertWarns(RuntimeWarning):
+            # Will raise a warning because never awaited
+            mock.async_method()
+        mock.async_method.assert_called()
+        asyncio.run(self._await_coroutine(coroutine))
+        mock.async_method.assert_awaited_once()
+
+    def test_assert_called_once_and_awaited_twice(self):
+        mock = AsyncMock(AsyncClass)
+        coroutine = mock.async_method()
+        mock.async_method.assert_called_once()
+        asyncio.run(self._await_coroutine(coroutine))
+        with self.assertRaises(RuntimeError):
+            # Cannot reuse already awaited coroutine
+            asyncio.run(self._await_coroutine(coroutine))
+        mock.async_method.assert_awaited()
+
+    def test_assert_awaited_but_not_called(self):
+        with self.assertRaises(AssertionError):
+            self.mock.assert_awaited()
+        with self.assertRaises(AssertionError):
+            self.mock.assert_called()
+        with self.assertRaises(TypeError):
+            # You cannot await an AsyncMock, it must be a coroutine
+            asyncio.run(self._await_coroutine(self.mock))
+
+        with self.assertRaises(AssertionError):
+            self.mock.assert_awaited()
+        with self.assertRaises(AssertionError):
+            self.mock.assert_called()
+
+    def test_assert_has_calls_not_awaits(self):
+        kalls = [call('NormalFoo')]
+        with self.assertWarns(RuntimeWarning):
+            self.mock('NormalFoo')
+        self.mock.assert_has_calls(kalls)
+        with self.assertRaises(AssertionError):
+            self.mock.assert_has_awaits(kalls)
+
+    def test_assert_has_mock_calls_on_async_mock(self):
+        mock = AsyncMock()
+        with self.assertWarns(RuntimeWarning):
+            mock()
+        kalls_empty = [('', (), {})]
+        self.assertEqual(mock.mock_calls, kalls_empty)
+
+        kalls_args = ([call('NormalFoo'), call('baz')])
+        with self.assertWarns(RuntimeWarning):
+            self.mock('NormalFoo')
+            self.mock('baz')
+        self.assertEqual(self.mock.mock_calls, kalls_args)
+
+        a_mock = AsyncMock(AsyncClass)
+        with self.assertWarns(RuntimeWarning):
+            a_mock.async_method()
+        self.assertEqual(a_mock.async_method.mock_calls, kalls_empty)
+        self.assertEqual(a_mock.mock_calls, [call.async_method()])
+
+        kalls_args_kwargs = [call(), call(1, 2, 3, a=4, b=5)]
+        with self.assertWarns(RuntimeWarning):
+            a_mock.async_method(1, 2, 3, a=4, b=5)
+        self.assertEqual(a_mock.async_method.mock_calls, kalls_args_kwargs)
```

Comment:
IIUC in this test you are first getting the `mock`, then calling it to get the coro (and that raises a `RuntimeWarning`), then you test that the mock has been called with no args.  Then you call it two more times with two args (and that raises a warning again) and check that it's been called twice with two different args.  Then you do it again with an `async_method` with an empty call and a call with args.

If this is correct I would:
* add a comment about why the error is raised before the first `with`
* add a comment to separate the part that checks calling the mock to the part about calling the `async_method`
* move the definition of `kalls_args` and `kalls_args_kwargs` after the `with` block like you did in the first part.
* maybe split this test into two, since it's getting confused with `mock`, `self.mock` and `a_mock`
* at the end check that the kalls propagated to `a_mock.mock_calls` too?